assembly/src/parsers/io_ops.rs

```diff
@@ -256,69 +273,131 @@ pub fn parse_loc_storew(op: &Token) -> Result<Node, ParsingError> {
+    let param_value = match param_str.strip_prefix("0x") {
+        Some(param_str) => {
+            debug_assert!(param_str.len() <= HEX_CHUNK_SIZE);
+            parse_hex_param(op, param_idx, param_str)
         }
-        false
+        None => parse_decimal_param(op, param_idx),
     };
 
-    // parse the hex string into one or more valid field elements
-    if is_single_element {
-        // parse a single element in hex representation
-        let parsed_param = parse_hex_param(op, param_idx, param_str)?;
-        Ok(vec![parsed_param])
+    param_value.map_err(|_| ParsingError::invalid_param(op, param_idx))
+}
+
+/// Parses a hexadecimal parameter value into a u64 value valid for field element creation.
+fn parse_hex_param(op: &Token, param_idx: usize, param_str: &str) -> Result<u64, ParsingError> {
+    debug_assert_eq!(param_str.len() % 2, 0, "hexadecimal string should be even");
+    match u64::from_str_radix(param_str, 16) {
+        Ok(value) => {
+            if value >= Felt::MODULUS {
+                return Err(ParsingError::invalid_param_with_reason(
+                    op,
+                    param_idx,
+                    format!("parameter value must be smaller than {}", Felt::MODULUS).as_str(),
+                ));
+            }
+            Ok(value)
+        }
+        Err(_) => Err(ParsingError::invalid_param(op, param_idx)),
+    }
+}
+
+/// Parses a decimal parameter value into a u64 value valid for field element creation.
+fn parse_decimal_param(op: &Token, param_idx: usize) -> Result<u64, ParsingError> {
+    let value = parse_param::<u64>(op, param_idx)?;
+    if value >= Felt::MODULUS {
+        return Err(ParsingError::invalid_param_with_reason(
+            op,
+            param_idx,
+            format!("parameter value must be smaller than {}", Felt::MODULUS).as_str(),
+        ));
+    }
+    Ok(value)
+}
+
+/// Determines the minimal type appropriate for provided value and returns appropriate instruction
+/// for this value
+fn build_push_one_instruction(value: u64) -> Result<Node, ParsingError> {
+    if value <= u8::MAX as u64 {
+        Ok(Instruction(PushU8(value as u8)))
+    } else if value <= u16::MAX as u64 {
+        Ok(Instruction(PushU16(value as u16)))
+    } else if value <= u32::MAX as u64 {
+        Ok(Instruction(PushU32(value as u32)))
+    } else if value < Felt::MODULUS {
+        Ok(Instruction(PushFelt(Felt::new(value))))
     } else {
-        // iterate over the multi-value hex string and parse each 8-byte chunk into a valid element
-        (0..param_str.len())
-            .step_by(HEX_CHUNK_SIZE)
-            .map(|i| parse_hex_param(op, param_idx, &param_str[i..i + HEX_CHUNK_SIZE]))
-            .collect()
+        unreachable!()
+    }
+}
+
+/// Determines the minimal type appropriate for provided values and returns appropriate instruction
+/// for this values
+fn build_push_many_instruction(values: Vec<u64>) -> Result<Node, ParsingError> {
```

Comment:
It looks like we are taking an iterator of `u64` that also implements `ExactSizeIterator` instead of owned vec. If we ask for owned vec, we force the user to realloc the values every time when we actually just take references and transform it into the appropriate type (u8, u16, etc)

Reply:
I'm not sure I understand what do you mean. Are you talking about process of creating a new vec of appropriate type (`values.iter().map(...)..`)?  That is, there is an opportunity not to force users to reallocate values because we just use references, right?

Reply:
Every access to `value` within this scope is to create a `.iter()`, meaning we should instead take an iterator as argument of the function

Reply:
Indeed! Thank you!